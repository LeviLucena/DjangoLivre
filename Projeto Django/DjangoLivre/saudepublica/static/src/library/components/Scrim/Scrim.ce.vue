<template>
  <div
    v-show="isActive"
    :id="id"
    ref="scrim"
    class="br-scrim foco"
    :class="isActive ? 'active' : null"
    :show="isActive"
    :data-visible="isActive"
    :aria-expanded="isActive"
    @click="cancel()"
  >
    <div
      ref="content"
      :class="contentClass"
      :tabindex="autofocusContent ? 0 : -1"
      @keydown.stop="handleKeyDown"
      @click.stop
    >
      <slot></slot>
    </div>
  </div>
</template>

<script>
const Scrim = {
  name: "br-scrim",
  props: {
    /** **[OPCIONAL]** Id do componente scrim. */
    id: {
      type: String,
      default: null,
    },
    /** **[OPCIONAL]** Ativa/desativa o scrim. */
    show: {
      type: Boolean,
      default: false,
    },
    /** **[OPCIONAL]** Centraliza o conteúdo dentro do scrim. */
    centerContent: {
      type: Boolean,
      default: false,
    },
    /** **[OPCIONAL]** Define se o conteúdo deve receber foco ao ser acionado. */
    autofocusContent: {
      type: Boolean,
      default: false,
    },
    /** **[OPCIONAL]** Desativa o fechamento do scrim ao ser clicado. */
    disableCloseOnClick: {
      type: Boolean,
      default: false,
    },
  },
  emits: ["update:show", "hide"],
  data() {
    return {
      isActive: this.show,
    }
  },
  watch: {
    show(value) {
      this.isActive = value
      this.$emit("update:show", value)
    },
  },
  mounted() {
    document.addEventListener("keyup", this.keyPress)
    if (this.autofocusContent) {
      this.$nextTick(() => {
        this.$refs.content.focus()
      })
    }
  },
  computed: {
    contentClass() {
      return {
        "br-modal": this.centerContent,
      }
    },
  },
  methods: {
    cancel() {
      if (!this.disableCloseOnClick) {
        this.isActive = !this.isActive
        this.hide()
      }
    },
    hide() {
      this.$emit("hide", "true")
      this.$emit("update:show", false)
    },
    keyPress(event) {
      if (event.key === "Escape" || event.key === "27") this.cancel()
    },
    handleKeyDown(event) {
      if (this.isActive && event.target === this.$refs.content && this.autofocusContent) {
        this.keyPress()
        event.stopPropagation()
      }
    },
  },
}

export default Scrim
</script>

<style lang="scss">
@import "~@govbr-ds/core/src/partial/scss/_base";
@import "~@govbr-ds/core/src/partial/scss/_utilities";
@import "~@govbr-ds/core/src/components/scrim/_scrim";

.br-modal {
  outline: none;
}
</style>
