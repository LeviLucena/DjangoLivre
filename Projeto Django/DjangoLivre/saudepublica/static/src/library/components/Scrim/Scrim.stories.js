import { action } from '@storybook/addon-actions'
import { kebabiseArgs } from '../../util/Utils.js'
import BrScrim from './Scrim.ce.vue'

const slotDefault = `
<div class="p-3x">
<br-input
  label="Label / Rótulo"
  placeholder="Placeholder"
></br-input>

</div>
`

export default {
  title: 'Componentes/Scrim',
  component: BrScrim,
  argTypes: {
    id: {
      control: 'text',
    },
    show: {
      control: 'boolean',
    },
    default: {
      control: 'text',
      description:
        '**[OBRIGATÓRIO]** Elemento que surge na tela juntamente com o Scrim, que deve ser passado por slot.',
      table: {
        type: {
          summary: 'string | Element',
        },
      },
    },
  },
}

const Template = (args) => ({
  props: Object.keys(args),
  setup() {
    return { args: kebabiseArgs(args) }
  },
  template: `

  <br-scrim v-bind="args" @hide="handleEvent" @update:show="handleShow" >
      ${args.default}
  </br-scrim>
  `,
  methods: { handleEvent: action('hide'), handleShow: action('update:show') },
})

export const Padrão = Template.bind({})
Padrão.args = {
  id: 'scrimexample',
  show: true,
  default: slotDefault,
  centerContent: true,
  autofocusContent: true,
}
