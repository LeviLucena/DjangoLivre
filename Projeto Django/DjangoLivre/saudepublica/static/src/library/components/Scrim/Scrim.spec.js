import { mount } from '@vue/test-utils'
import Scrim from './Scrim.ce.vue'

describe('Scrim', () => {
  it('renders when show is true', async () => {
    const wrapper = mount(Scrim, {
      props: {
        show: true,
      },
    })
    expect(wrapper.isVisible()).toBe(true)
    expect(wrapper.find('.br-scrim.active').exists()).toBe(true)
  })

  it('does not render when show is false', async () => {
    const wrapper = mount(Scrim, {
      props: {
        show: false,
      },
    })
    expect(wrapper.isVisible()).toBe(false)
    expect(wrapper.find('.br-scrim.active').exists()).toBe(false)
  })

  it('emits hide event when canceled', async () => {
    const wrapper = mount(Scrim, {
      props: {
        show: true,
      },
    })
    await wrapper.find('.br-scrim.active').trigger('click')
    expect(wrapper.emitted('hide')).toBeTruthy()
  })

  it('does not hide when disableCloseOnClick is true', async () => {
    const wrapper = mount(Scrim, {
      props: {
        show: true,
        disableCloseOnClick: true,
      },
    })
    wrapper.find('.br-scrim.active').trigger('click')
    expect(wrapper.emitted('hide')).toBeFalsy()
  })

  it('closes when Esc key is pressed', async () => {
    const wrapper = mount(Scrim, {
      props: {
        isActive: true,
      },
    })
    wrapper.vm.$nextTick(() => {
      wrapper.vm.keyPress({ key: 'Escape' })
      expect(wrapper.emitted('hide')).toBeTruthy()
    })
  })

  it('closes when "27" key is pressed', async () => {
    const wrapper = mount(Scrim, {
      props: {
        isActive: true,
      },
    })
    wrapper.vm.$nextTick(() => {
      wrapper.vm.keyPress({ key: '27' })
      expect(wrapper.emitted('hide')).toBeTruthy()
    })
  })

  it('autofocusContent should set focus on content on mount', () => {
    const wrapper = mount(Scrim, {
      props: {
        show: true,
        autofocusContent: true,
      },
      slots: {
        default: '<div id="content">Scrim Content</div>',
      },
    })
    expect(wrapper.find('[tabindex="0"]').exists()).toBe(true)
  })

  it('autofocusContent false should not set focus on content on mount', () => {
    const wrapper = mount(Scrim, {
      props: {
        show: true,
        autofocusContent: false,
      },
      slots: {
        default: '<div id="content">Scrim Content</div>',
      },
    })
    expect(wrapper.find('[tabindex="-1"]').exists()).toBe(true)
    expect(wrapper.element).toMatchSnapshot()
  })

  it('centerContent should center the content', () => {
    const wrapper = mount(Scrim, {
      props: {
        show: true,
        centerContent: true,
      },
      slots: {
        default: '<div id="content">Scrim Content</div>',
      },
    })
    expect(wrapper.find('.br-modal').exists()).toBe(true)
  })
})
