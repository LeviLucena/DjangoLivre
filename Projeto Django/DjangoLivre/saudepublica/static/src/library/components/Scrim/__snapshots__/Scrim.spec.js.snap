// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrim autofocusContent false should not set focus on content on mount 1`] = `
<div
  aria-expanded="true"
  class="br-scrim foco active"
  data-visible="true"
  show="true"
>
  <div
    class=""
    tabindex="-1"
  >
    
    <div
      id="content"
    >
      Scrim Content
    </div>
    
  </div>
</div>
`;
